File: JavaScript/454. 4Sum II.js
var fourSumCount = function(A, B, C, D) {
    let dict = {};
    
    for (let i = 0; i < A.length; i++){
        for (let j = 0; j < B.length; j++){
            if (!((A[i] + B[j]) in dict)) {
                dict[(A[i] + B[j])] = 0;
            }
            
            dict[(A[i] + B[j])] += 1;
        }
    }
    
    let ans = 0;
    for (let i = 0; i < C.length; i++){
        for (let j = 0; j < D.length; j++){
            if (-(C[i] + D[j]) in dict) {
                ans += dict[-(C[i] + D[j])];
            }
        }
    }
    return ans;
    
};

// hashtable solution 
// 

File: Leetcode First 200 Python solution/454. 4Sum II.py
# 4 sum II
class Solution:
    def fourSumCount(self, A, B, C, D):
        
        hashtable = collections.defaultdict(int)
        
        for a in A:
            for b in B:
                hashtable[a + b] +=1
        
        ans = 0
        
        for c in C:
            for d in D:
                ans += hashtable[-c - d]
        
        return ans

# o(n^2) solution here